source_files/gpu.c

```diff
@@ -38,98 +38,103 @@
 
 #include "gpu.h"
 
-#define CUDA_ERROR_CHECK
-#define CUDA_SAFE_CALL( cuerr ) cuda_safe_call( cuerr, __FILE__, __LINE__ )
+#define CUDA_SAFE_CALL( cuerr, dev_index ) cuda_safe_call( cuerr, dev_index, __FILE__, __LINE__ )
 
-static void *A,*B;
-static unsigned int size,useDouble,useDevice,verbose; //matrixsize,switch for double precision and numbers of GPUs to use.
-static gpustruct * gpuvar=NULL;
+static volatile gpustruct_t * gpuvar;
 
 //CUDA Error Checking
-static inline void cuda_safe_call( cudaError_t cuerr, const char *file, const int line )
-{
-#ifdef CUDA_ERROR_CHECK
+static inline void cuda_safe_call( cudaError_t cuerr, int dev_index, const char *file, const int line ) {
     if ( cuerr != cudaSuccess)
     {
-        fprintf( stderr, "    - CUDA_SAFE_CALL() failed at %s:%i : %i %s\n",
-                 file, line, cuerr,cudaGetErrorString( cuerr ) );
+        fprintf( stderr, "    - CUDA error at %s:%i : error code = %i(%s), device index:%i\n",
+                 file, line, cuerr, cudaGetErrorString( cuerr ), dev_index );
         exit(cuerr);
     }
-#endif
 
     return;
 }
 
-int roundUp(int numToRound, int multiple)  
-{  
- if(multiple == 0)  
- {  
-  return numToRound;  
- }  
+static int roundUp(int numToRound, int multiple) {  
+    if(multiple == 0) {  
+        return numToRound;
+    }  
 
- int remainder = numToRound % multiple; 
- if (remainder == 0)
-  {
-    return numToRound; 
-  }
+    int remainder = numToRound % multiple; 
+    if (remainder == 0) {
+        return numToRound;
+    }
 
- return numToRound + multiple - remainder; 
+    return numToRound + multiple - remainder; 
 } 
 
-int ipow(int base,int exp) {
+static int ipow(int base,int exp) {
     int result = 1;
     while (exp) {
         if (exp & 1) result *= base;
         exp >>= 1;
         base *= base;
     }
+
     return result;
 }
 
 static int bbs(void) {
     static int s;
```

Comment:
s and runned are used concurrently. probably better to pass them to the functions within each thread, initialized by 0 and 123,
i.e. 
-static int ipow(int base,int exp) --> -int ipow(int base,int exp, int * s ,int *runned)
 static int bbs(void) -->  static int bbs(int * s ,int *runned)
static void* fillup(void* array, int useD, int size) --> static void* fillup(void* array, int useD, int size, int * s ,int *runned)